crates/ruff_linter/src/rules/refurb/rules/isinstance_type_none.rs

```diff
@@ -88,29 +85,81 @@ fn is_none(expr: &Expr) -> bool {
             Expr::Call(ast::ExprCall {
                 func, arguments, ..
-            }) if arguments.len() == 1 => {
-                if let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() {
-                    if id.as_str() == "type" {
-                        return matches!(arguments.args.first(), Some(Expr::NoneLiteral(_)));
-                    }
+            }) => {
+                if !semantic.match_builtin_expr(func, "type") {
+                    return false;
+                }
+
+                if !arguments.keywords.is_empty() {
+                    return false;
                 }
-                false
+
+                matches!(arguments.args.as_ref(), [Expr::NoneLiteral(_)])
             }
 
             // Ex) `(type(None),)`
-            Expr::Tuple(tuple) => tuple.iter().all(|element| inner(element, false)),
+            Expr::Tuple(tuple) => tuple.iter().all(|element| inner(element, false, semantic)),
 
             // Ex) `type(None) | type(None)`
             Expr::BinOp(ast::ExprBinOp {
                 left,
                 op: Operator::BitOr,
                 right,
                 ..
-            }) => inner(left, true) && inner(right, true),
+            }) => {
+                // `None | None` is a `TypeError` at runtime
+                if left.is_none_literal_expr() && right.is_none_literal_expr() {
+                    return false;
+                }
+
+                inner(left, true, semantic) && inner(right, true, semantic)
+            }
+
+            // Ex) `Union[None, ...]`
+            Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+                if !semantic.match_typing_expr(value, "Union") {
+                    return false;
+                }
+
+                match slice.as_ref() {
+                    Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+                        elts.iter().all(|element| inner(element, true, semantic))
+                    }
+                    slice => inner(slice, true, semantic),
+                }
+            }
 
             // Otherwise, return false.
             _ => false,
         }
     }
-    inner(expr, false)
+    inner(expr, false, semantic)
+}
+
+fn replace_with_identity_check(expr: &Expr, range: TextRange, checker: &Checker) -> Fix {
+    let (semantic, generator) = (checker.semantic(), checker.generator());
+
+    let new_expr = Expr::Compare(ast::ExprCompare {
+        left: expr.clone().into(),
+        ops: [CmpOp::Is].into(),
+        comparators: [ast::ExprNoneLiteral::default().into()].into(),
+        range: TextRange::default(),
+    });
+
+    let new_content = generator.expr(&new_expr);
+    let new_content = if semantic.current_expression_parent().is_some() {
+        format!("({new_content})")
+    } else {
+        new_content
+    };
+
+    let applicability = if checker.comment_ranges().intersects(range) {
+        Applicability::Unsafe
+    } else {
+        Applicability::Safe
+    };
```

Comment:
Please could you add a `## Fix safety` section to the docs stating why the fix is sometimes marked as unsafe?